group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,240 @@
+                    }
+                }
+
+                // Number of partitions left to reach the minRequiredQuota.
+                int remaining = minRequiredQuota - currentAssignmentSize;
+
+                // There are 3 cases w.r.t value of remaining
+                // 1) remaining < 0 this means that the member has more than the min required amount.
+                if (remaining < 0 && numMembersWithExtraPartition > 0) {
+                    // In order to remain as sticky as possible, since the order of members can be different, we want the members that already had extra
+                    // partitions to retain them if it's still required, instead of assigning the extras to the first few members directly.
+                    numMembersWithExtraPartition--;
+                    // Since we already added the minimumRequiredQuota of partitions in the previous step (until minReq - 1), we just need to
+                    // add the extra partition that will be present at the index right after min quota was satisfied.
+                    assignedStickyPartitionsPerTopic.computeIfAbsent(topicId, k -> new HashSet<>()).add(currentAssignmentListForTopic.get(minRequiredQuota));
+                    membersWithNewAssignmentPerTopic.computeIfAbsent(memberId, k -> new HashMap<>()).computeIfAbsent(topicId, k -> new HashSet<>()).add(currentAssignmentListForTopic.get(minRequiredQuota));
+                } else {
+                    // 2) If remaining = 0 it has min req partitions but there is scope for getting an extra partition later on, so it is a potentialUnfilledMember.
+                    // 3) If remaining > 0 it doesn't even have the min required partitions, and it definitely is unfilled, so it should be added to potentialUnfilledMembers.
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    potentiallyUnfilledMembers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // If remaining > 0 after increasing the required quota due to the extra partition, add potentially unfilled member to the unfilled members list.
+            for (RemainingAssignmentsForMember<String, Integer> pair : potentiallyUnfilledMembers) {
+                String memberId = pair.memberId();
+                Integer remaining = pair.remaining();
+                if (numMembersWithExtraPartition > 0) {
+                    remaining++;
+                    numMembersWithExtraPartition--;
+                }
+                if (remaining > 0) {
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    unfilledMembersPerTopic.computeIfAbsent(topicId, k -> new ArrayList<>()).add(newPair);
+                }
+            }
+        });
+
+        // Step 4
```

Comment:
Is there a reason why we don't do step 4 and 5 directly in the `membersPerTopic.forEach((topicId, membersForTopic)` loop?

Reply:
like calculate unassigned partitions and assign them at the same time?

Reply:
It does not have to be at the same time. I was wondering if there a reason why we need to do step 4 and 5 afterwards with all the unfilled members vs doing it per topic right after step 3.

Reply:
We could've computed if the partition is still unassigned in step 5 directly and assigned it but since this is a range assignor I need all the available partitions in in a sorted list and the list provided is sorted since we iterate through 0-n where n is the total partitions and only add it to unassigned list iff it doesn't exist in the sticky partitions list

Reply:
I hope I understood the question correctly

Reply:
Also the entire unfilled members per topic list needs to be populated since the ranges depend on how many partitions were assigned to the prev member